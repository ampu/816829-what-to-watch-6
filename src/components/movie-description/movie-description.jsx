import React from 'react';
import PropTypes from 'prop-types';
import {generatePath, useHistory} from 'react-router-dom';
import {connect} from 'react-redux';

import {MainPath} from '../../constants/paths';
import OperationStatus from '../../constants/operation-status';

import PlayButton from '../play-button/play-button';
import MyListButton from '../my-list-button/my-list-button';

import movieType from '../../typings/movie-type';
import {selectLoginStatus} from '../../store/selectors/user-selectors';

import operationStatusType from '../../typings/operation-status-type';

const MovieDescription = ({loginStatus, movie = {}, children}) => {
  const {
    title = ``,
    genre = ``,
    year = ``,
  } = movie;

  const history = useHistory();

  const handlePlayClick = (evt) => {
    evt.preventDefault();
    history.push(generatePath(MainPath.PLAYER, movie));
  };

  return title && (
    <div className="movie-card__desc">
      <h2 className="movie-card__title">{title}</h2>
      <p className="movie-card__meta">
        {genre && <span className="movie-card__genre">{genre}</span>}
        {year && <span className="movie-card__year">{year}</span>}
      </p>

      <div className="movie-card__buttons">
        <PlayButton className="btn btn--play movie-card__button" onClick={handlePlayClick}/>

        {loginStatus === OperationStatus.RESOLVED && (
          <MyListButton className="btn btn--list movie-card__button" movie={movie}/>
        )}

        {children}
      </div>
    </div>
  );
};

MovieDescription.propTypes = {
  loginStatus: operationStatusType,
  movie: PropTypes.shape({
    title: movieType.title,
    genre: movieType.genre,
    year: movieType.year,
  }),
  children: PropTypes.any,
};

const mapStateToProps = (state) => ({
  loginStatus: selectLoginStatus(state),
});

export {MovieDescription};
export default connect(mapStateToProps)(MovieDescription);
