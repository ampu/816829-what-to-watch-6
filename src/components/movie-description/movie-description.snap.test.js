import React from 'react';

import {render} from '@testing-library/react';

import OperationStatus from '../../constants/operation-status';

import {MovieDescription} from './movie-description';

describe(`MovieDescription component`, () => {
  it(`Should render according to the snapshot with only movie id provider`, () => {
    Object.values(OperationStatus).forEach((loginStatus) => {
      if (loginStatus !== OperationStatus.RESOLVED) {
        const movie = {
          id: `fakeId`,
        };
        const {container} = render(<MovieDescription loginStatus={loginStatus} movie={movie}/>);
        expect(container).toMatchSnapshot();
      }
    });
  });

  it(`Should render according to the snapshot with extra information provider`, () => {
    Object.values(OperationStatus).forEach((loginStatus) => {
      if (loginStatus !== OperationStatus.RESOLVED) {
        const movie = {
          id: `fakeId`,
          title: `fakeTitle`,
          genre: `fakeGenre`,
          year: 2021,
        };
        const {container} = render((
          <MovieDescription loginStatus={loginStatus} movie={movie}>
            fakeChildren
          </MovieDescription>
        ));
        expect(container).toMatchSnapshot();
      }
    });
  });
});
