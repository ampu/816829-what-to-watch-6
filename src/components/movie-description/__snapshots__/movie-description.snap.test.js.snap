// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDescription component Should render according to the snapshot with extra information provider 1`] = `
<div>
  <div
    class="movie-card__desc"
  >
    <h2
      class="movie-card__title"
    >
      fakeTitle
    </h2>
    <p
      class="movie-card__meta"
    >
      <span
        class="movie-card__genre"
      >
        fakeGenre
      </span>
      <span
        class="movie-card__year"
      >
        2021
      </span>
    </p>
    <div
      class="movie-card__buttons"
    >
      <button
        class="btn btn--play movie-card__button"
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlink:href="#play-s"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
      fakeChildren
    </div>
  </div>
</div>
`;

exports[`MovieDescription component Should render according to the snapshot with extra information provider 2`] = `
<div>
  <div
    class="movie-card__desc"
  >
    <h2
      class="movie-card__title"
    >
      fakeTitle
    </h2>
    <p
      class="movie-card__meta"
    >
      <span
        class="movie-card__genre"
      >
        fakeGenre
      </span>
      <span
        class="movie-card__year"
      >
        2021
      </span>
    </p>
    <div
      class="movie-card__buttons"
    >
      <button
        class="btn btn--play movie-card__button"
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlink:href="#play-s"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
      fakeChildren
    </div>
  </div>
</div>
`;

exports[`MovieDescription component Should render according to the snapshot with extra information provider 3`] = `
<div>
  <div
    class="movie-card__desc"
  >
    <h2
      class="movie-card__title"
    >
      fakeTitle
    </h2>
    <p
      class="movie-card__meta"
    >
      <span
        class="movie-card__genre"
      >
        fakeGenre
      </span>
      <span
        class="movie-card__year"
      >
        2021
      </span>
    </p>
    <div
      class="movie-card__buttons"
    >
      <button
        class="btn btn--play movie-card__button"
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlink:href="#play-s"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
      fakeChildren
    </div>
  </div>
</div>
`;

exports[`MovieDescription component Should render according to the snapshot with only movie id provider 1`] = `
<div>
  <div
    class="movie-card__desc"
  >
    <h2
      class="movie-card__title"
    />
    <p
      class="movie-card__meta"
    >
      
      
    </p>
    <div
      class="movie-card__buttons"
    >
      <button
        class="btn btn--play movie-card__button"
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlink:href="#play-s"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`MovieDescription component Should render according to the snapshot with only movie id provider 2`] = `
<div>
  <div
    class="movie-card__desc"
  >
    <h2
      class="movie-card__title"
    />
    <p
      class="movie-card__meta"
    >
      
      
    </p>
    <div
      class="movie-card__buttons"
    >
      <button
        class="btn btn--play movie-card__button"
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlink:href="#play-s"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`MovieDescription component Should render according to the snapshot with only movie id provider 3`] = `
<div>
  <div
    class="movie-card__desc"
  >
    <h2
      class="movie-card__title"
    />
    <p
      class="movie-card__meta"
    >
      
      
    </p>
    <div
      class="movie-card__buttons"
    >
      <button
        class="btn btn--play movie-card__button"
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlink:href="#play-s"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
    </div>
  </div>
</div>
`;
